vision_transformer/main.py

```diff
@@ -0,0 +1,244 @@
+        pos_embed = self.pos_embedding[:, :n + 1, :]
+        linear_projection += pos_embed
+
+        return linear_projection
+
+
+class EncoderBlock(nn.Module):
+
+    def __init__(self, args):
+        super(EncoderBlock, self).__init__()
+
+        self.latent_size = args.latent_size
+        self.num_heads = args.num_heads
+        use_cuda = not args.no_cuda and torch.cuda.is_available()
+        self.device = torch.device("cuda" if use_cuda else "cpu")
+        self.dropout = args.dropout
+
+        self.norm = nn.LayerNorm(self.latent_size)
+        self.attention = nn.MultiheadAttention(self.latent_size, self.num_heads, dropout=self.dropout)
+        self.enc_MLP = nn.Sequential(
+            nn.Linear(self.latent_size, self.latent_size * 4),
+            nn.GELU(),
+            nn.Dropout(self.dropout),
+            nn.Linear(self.latent_size * 4, self.latent_size),
+            nn.Dropout(self.dropout)
+        )
+
+    def forward(self, emb_patches):
+        first_norm = self.norm(emb_patches)
+        attention_out = self.attention(first_norm, first_norm, first_norm)[0]
+        first_added = attention_out + emb_patches
+        second_norm = self.norm(first_added)
+        mlp_out = self.enc_MLP(second_norm)
+        output = mlp_out + first_added
+
+        return output
+
+
+class ViT(nn.Module):
+    def __init__(self, args):
+        super(ViT, self).__init__()
+
+        self.num_encoders = args.num_encoders
+        self.latent_size = args.latent_size
+        use_cuda = not args.no_cuda and torch.cuda.is_available()
+        self.device = torch.device("cuda" if use_cuda else "cpu")
+        self.num_classes = args.num_classes
+        self.dropout = args.dropout
+
+        self.embedding = InputEmbedding(args)
+        # Encoder Stack
+        self.encoders = nn.ModuleList([EncoderBlock(args) for _ in range(self.num_encoders)])
+        self.MLPHead = nn.Sequential(
+            nn.LayerNorm(self.latent_size),
+            nn.Linear(self.latent_size, self.latent_size),
+            nn.Linear(self.latent_size, self.num_classes),
+        )
+
+    def forward(self, test_input):
+        enc_output = self.embedding(test_input)
+        for enc_layer in self.encoders:
+            enc_output = enc_layer(enc_output)
+
+        class_token_embed = enc_output[:, 0]
+        return self.MLPHead(class_token_embed)
+
+
+class TrainEval:
+
+    def __init__(self, model, train_dataloader, val_dataloader, optimizer, criterion, epochs, device):
+        self.model = model
+        self.train_dataloader = train_dataloader
+        self.val_dataloader = val_dataloader
+        self.optimizer = optimizer
+        self.criterion = criterion
+        self.epoch = epochs
+        self.device = device
+
+    def train_fn(self, current_epoch):
+        self.model.train()
+        total_loss = 0.0
+        tk = tqdm(self.train_dataloader, desc="EPOCH" + "[TRAIN]" + str(current_epoch + 1) + "/" + str(self.epoch))
```

Comment:
just triggered CI let's see if tqdm is supported there, otherwise we have a fake tqdm `from torch.hub import tqdm` that's also useful for a basic progress bar